### PR TITLE
feat: [CDS-52537]: Http step delegate logging

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes
1. Moved Http module from 960-api-services to 930-delegate-tasks to leverage delegate logging
2. Fixed BUILD.bazel dependencies 
3. Open log stream before building task request on pms side
4. Access stream on the delegate side and push logs to same log key.
5. Access stream while handling task result and push logs to same log key. 
6. Close stream finally when step finishes execution, step is aborted.

image
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)

### CODE REVIEW

**Comment by hingerabhinav** [comment]
trigger ti0, ti1, ti2, ti3, ti4

**Review by vaibhav-si — DISMISSED** [review]
> Reviewed 4 of 7 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: 11 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Comment by yogesh-chauhan** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java` line 35 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyMbe74ZhQf4wkxXLM:-NOyMbe74ZhQf4wkxXLN:b-aotq46) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java#L35)):*
> ```Java
> @Slf4j
> public class HttpTaskNG extends AbstractDelegateRunnableTask {
>   public static final String COMMAND_UNIT = "Execute";
> ```

does this need to be public ?

**Comment by yogesh-chauhan** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java` line 55 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyMiskDLTglEluYZSp:-NOyMisl2ZLyE-bS0K6r:biju5e2) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java#L55)):*
> ```Java
>     CommandUnitsProgress commandUnitsProgress = CommandUnitsProgress.builder().build();
>     // log stream opened already
>     LogCallback executionLogCallback =
> ```

can these be final ?

**Comment by yogesh-chauhan** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java` line 86 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyMwlo9Ae6c78GZ8Kl:-NOyMwlo9Ae6c78GZ8Km:b-1hwa6b) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java#L86)):*
> ```Java
> public class HttpServiceImpl implements HttpService {
>   @Override
>   public HttpInternalResponse executeUrl(HttpInternalConfig config) throws IOException {
> ```

can we pass "io.harness.logging.NoopExecutionCallback" instead of null ?

**Comment by yogesh-chauhan** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 128 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyNPCn1n3l7Ni4q1uv:-NOyNPCn1n3l7Ni4q1uw:b-1jhezj) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L128)):*
> ```Java
>     //    return TaskRequestsUtils.prepareTaskRequestWithTaskSelector(ambiance, taskData, referenceFalseKryoSerializer,
>     //        TaskSelectorYaml.toTaskSelector(httpStepParameters.delegateSelectors));
> ```

nit: remove this commented out code

**Comment by yogesh-chauhan** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 186 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyNdg_DzeVdtPtdjPf:-NOyNdg_DzeVdtPtdjPg:b-tq5eih) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L186)):*
> ```Java
>   private void closeLogStream(Ambiance ambiance) {
>     try {
>       Thread.sleep(500, 0);
> ```

why do we need a sleep here ?

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 4 of 7 files at r1, 5 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: 2 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Comment by yogesh-chauhan** [comment]
Can you summarize how we are going to release these changes since both pipeline service and delegate side logic is impacted by these ? Do you see any backward compatibility risks here

**Comment by hingerabhinav** [comment]
> Can you summarize how we are going to release these changes since both pipeline service and delegate side logic is impacted by these ? Do you see any backward compatibility risks here

I think major risk is the log key. Pipeline service opens a log stream which delegate accesses and finally pipeline service accesses again to write assertion logs. 

new PMS - new DEL / old PMS - old DEL -> works 
new PMS - old DEL -> PMS opens a stream and writes to it. Delegate won't do anything. Same logs as before.
old PMS - new DEL -> no stream opened by PMS and Delegate tries to write to it but can't. Then PMS opens a log stream to write assertion logs. Same logs as before.

I will test out these scenarios and update here. What do you think?

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: 3 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

*[a discussion](https://reviewable.io/reviews/harness/harness-core/44229#-NOyNkXS85WjEQjTd4Aa:-NP2-CR_7U6RHxmMhzLz:b-gq9ggr) (no related file):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Can you summarize how we are going to release these changes since both pipeline service and delegate side logic is impacted by these ? Do you see any backward compatibility risks here
</blockquote></details>

Done. I have verified older delegate with newer pipeline service, we get the old behavior when delegate is old.

___
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java` line 35 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyMbe74ZhQf4wkxXLM:-NP2-M269oD-ePv_WQoJ:bs0ynke) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java#L35)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

does this need to be public ?
</blockquote></details>

Yes. This is used by PMS to open stream with this key.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java` line 55 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyMiskDLTglEluYZSp:-NP2-Pv56tEayRHvL6kW:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/930-delegate-tasks/src/main/java/io/harness/delegate/http/HttpTaskNG.java#L55)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can these be final ?
</blockquote></details>

Done
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 186 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyNdg_DzeVdtPtdjPf:-NP2-Y8V2EmAvpmbKwlc:buwhece) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L186)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

why do we need a sleep here ?
</blockquote></details>

since log service pushes logs in batches. we're giving some time for logs to be pushed.
___
*[`930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java` line 86 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NOyMwlo9Ae6c78GZ8Kl:-NP2-UHR8nxAUWxNSNJK:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java#L86)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we pass "io.harness.logging.NoopExecutionCallback" instead of null ?
</blockquote></details>

Done

**Comment by yogesh-chauhan** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 61 at r3](https://reviewable.io/reviews/harness/harness-core/44229#-NP2fpsfE-GELTO1MITi:-NP2fpsfE-GELTO1MITj:bxi21ow) ([raw file](https://github.com/harness/harness-core/blob/163c8932494fb9225d3a1962993b038ab8e06d90/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L61)):*
> ```Java
> import com.google.inject.Inject;
> import com.google.inject.name.Named;
> import java.util.*;
> ```

we should not do wildcard imports

**Comment by yogesh-chauhan** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java` line 171 at r2](https://reviewable.io/reviews/harness/harness-core/44229#-NP2fuTIBuTj96XoHV0O:-NP2fuTIBuTj96XoHV0P:b-fo1a3w) ([raw file](https://github.com/harness/harness-core/blob/78f5439094e5db8dc6913596b47f97d5fec090dd/930-delegate-tasks/src/main/java/io/harness/http/HttpServiceImpl.java#L171)):*
> ```Java
> 
>   private void saveLogs(LogCallback executionLogCallback, String message) {
>     if (executionLogCallback != null) {
> ```

null check is good to have

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 3 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: 4 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 4 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: 5 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Review by vaibhav-si — APPROVED** [review]
> Reviewed 1 of 4 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: Waiting on team(s): del-team.

**Review by vaibhav-si — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: Waiting on team(s): del-team. 

___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 187 at r4](https://reviewable.io/reviews/harness/harness-core/44229#-NPC7rzNAxO-QdLwO81n:-NPC7rzNAxO-QdLwO81o:b-mipix9) ([raw file](https://github.com/harness/harness-core/blob/d1888321f538dec86657fd4ad4eaaba955bd9236/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L187)):*
> ```Java
>   private void closeLogStream(Ambiance ambiance) {
>     try {
>       Thread.sleep(500, 0);
> ```

Delegate should flush the logs before completing the task IMO. This synchronization should not be needed. For now, it's okay but we can create a ticket for them to look at.

**Comment by hingerabhinav** [comment]
> Delegate should flush the logs before completing the task IMO. This synchronization should not be needed. For now, it's okay but we can create a ticket for them to look at.

Sure @vaibhav-si will create a ticket for this

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 4 of 7 files at r1, 2 of 7 files at r2, 1 of 4 files at r3, 2 of 4 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: 3 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Review by vaibhav-si — APPROVED** [review]
> Reviewed 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44229)* status: :shipit: complete! ✓
